### PR TITLE
Add the function to compare yaml

### PR DESCRIPTION
**Pull request by dd-jy**
### Add the function to compare yaml.

Result of  tests/test_compare_yaml.py : 
```
{'add': [['new-test-oss', [['1.0.1', 'Apache-2.0,MIT']]]], 'delete': [['del-test-oss', [['1.0.1', 'Apache-2.0,MIT']]]], 'update': [['add-license-oss', [['0.5.6', 'Apache-2.0,MPL-2.0']], [['0.5.3', 'Apache-2.0'], ['0.5.3', 'MPL-2.0']]], ['go:github.com/konsorten/go-windows-terminal-sequences', [['1.0.1', 'MIT']], [['1.0.1', 'Apache-2.0,MIT']]]]}
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Refactoring, Maintenance
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: src/fosslight_util/compare_yaml.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from fosslight_util.parsing_yaml import parsing_yml
+
+logger = logging.getLogger(LOGGER_NAME)
+_version = 'version'
+_license = 'license'
+_name = 'name'
+
+
+def compare_yaml(before_file, after_file):
+    before_oss_items, _ = parsing_yml(os.path.basename(before_file), os.path.dirname(before_file))
```
Comment (soimkim) [inline src/fosslight_util/compare_yaml.py:18]: parsing_yml(os.path.basename(before_file), os.path.dirname(before_file))
-> parsing_yml(before_file, os.path.dirname(before_file))

**Review by soimkim — CHANGES_REQUESTED** [review]

File: tests/test_compare_yaml.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# -*- coding: utf-8 -*-
+# Copyright (c) 2022 LG Electronics Inc.
+# SPDX-License-Identifier: Apache-2.0
+import os
+from fosslight_util.compare_yaml import compare_yaml
+
+
+def main():
+    cur = os.getcwd()
+    print(compare_yaml(os.path.join(cur, 'before.yaml'), os.path.join(cur, 'after.yaml')))
```
Comment (soimkim) [inline tests/test_compare_yaml.py:11]: (Minor) This is better for analyzing the results.
``` 
import json
result = compare_yaml(os.path.join(cur, 'before.yaml'), os.path.join(cur, 'after.yaml'))
print(json.dumps(result, indent=4, sort_keys=True))
```

**Comment by dd-jy** [comment]
@soimkim 
Thanks for reveiwing! I changed everything.

File: src/fosslight_util/compare_yaml.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    CHANGE = "change"
+    PREV = "prev"
+    NOW = "now"
+
+    comp_out = {ADD: [], DELETE: [], CHANGE: []}
+
+    b_name = list(set([bi[NAME] for bi in new_before]))
+    a_name = list(set([ai[NAME] for ai in after_items]))
+
+    removed_name = [bn for bn in b_name if bn not in a_name]
```
Comment (soimkim) [inline src/fosslight_util/compare_yaml.py:42]: [Minor] There seems to be no need to convert it to a list.
ex.
```
    b_name = set([bi[NAME] for bi in new_before])
    a_name = set([ai[NAME] for ai in after_items])
    removed_name = b_name - a_name
    changed_name = b_name ^ removed_name
    added_name = a_name ^ changed_name
```

Comment (soimkim) [reply src/fosslight_util/compare_yaml.py:42]: 1 more opinion.
Can't the following be changed to the difference set?
NOW: 
```    
changed_name = b_name ^ removed_name
added_name = a_name ^ changed_name
```
TO-BE:
```    
changed_name = b_name - removed_name
added_name = a_name - changed_name
```

File: src/fosslight_util/compare_yaml.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    changed_name = list(set(b_name) ^ set(removed_name))
+    added_name = list(set(a_name) ^ set(changed_name))
+
+    for bi in new_before:
+        if bi[NAME] in removed_name:
+            comp_out[DELETE].append(bi)
+        elif bi[NAME] in changed_name:
+            name_find = False
+            for ci in comp_out[CHANGE]:
```
Comment (soimkim) [inline src/fosslight_util/compare_yaml.py:51]: How about changing the inside of `elif bi[NAME] in changed_name` like this?
```
            name_find = False
            for ci in comp_out[CHANGE]:
                if ci[NAME] == bi[NAME]:
                    bi.pop(NAME)
                    ci[PREV].append(bi)
                    name_find = True
                    break

            if not name_find:
                item_info = {NAME: bi[NAME], PREV: [], NOW: []}
                bi.pop(NAME)
                item_info[PREV].append(bi)
                comp_out[CHANGE].append(item_info)
```
TO-BE:
```
            filtered = next(filter(lambda oss_dict: oss_dict['name'] == bi[NAME], comp_out[CHANGE]), None)
            if filtered:
                bi.pop(NAME)
                filtered[PREV].append(bi)
                name_find = True
            else:
                item_info = {NAME: bi[NAME], PREV: [], NOW: []}
                bi.pop(NAME)
                item_info[PREV].append(bi)
                comp_out[CHANGE].append(item_info)
```

File: src/fosslight_util/compare_yaml.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            if not name_find:
+                item_info = {NAME: bi[NAME], PREV: [], NOW: []}
+                bi.pop(NAME)
+                item_info[PREV].append(bi)
+                comp_out[CHANGE].append(item_info)
+
+    for ai in after_items:
+        if ai[NAME] in added_name:
+            comp_out[ADD].append(ai)
+        elif ai[NAME] in changed_name:
```
Comment (soimkim) [inline src/fosslight_util/compare_yaml.py:66]: How about changing the inside of `elif ai[NAME] in changed_name `like this?
AS-IS:
```
 for ci in comp_out[CHANGE]:
                if ci[NAME] == ai[NAME]:
                    ai.pop(NAME)
                    ci[NOW].append(ai)
                    break
```
TO-BE:
```
            filtered = next(filter(lambda oss_dict: oss_dict[NAME] == ai[NAME], comp_out[CHANGE]), None)
            if filtered:
                ai.pop(NAME)
                filtered[NOW].append(ai)
```

File: src/fosslight_util/compare_yaml.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    added_name = list(set(a_name) ^ set(changed_name))
+
+    for bi in new_before:
+        if bi[NAME] in removed_name:
+            comp_out[DELETE].append(bi)
+        elif bi[NAME] in changed_name:
+            name_find = False
+            for ci in comp_out[CHANGE]:
+                if ci[NAME] == bi[NAME]:
+                    bi.pop(NAME)
```
Comment (soimkim) [inline src/fosslight_util/compare_yaml.py:53]: [Question] Why is `bi.pop(NAME)` called?

Comment (soimkim) [reply src/fosslight_util/compare_yaml.py:53]: For printing!! 👍

**Pull request merged**